src/librustc/traits/select.rs

```diff
@@ -2612,6 +2617,37 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                     &[inner_target]));
             }
 
+            // (.., T) -> (.., U).
+            (&ty::TyTuple(tys_a, _), &ty::TyTuple(tys_b, _)) => {
```

Comment:
The code above `mk_adt` largely depends on def-ids and type parameters, so I doubt I can merge this part with tuple's. The codes below `mk_adt`/`mk_tup` are similar, but I don't think its worth merging...